### PR TITLE
fix: improve perf: switch parentUIDToChildren to map of sets, remove cache rebuild (#26863)

### LINKED ISSUE #26863: Application Controller Performance regression in 3.3.X
Checklist:

- [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
- [ ] I've included steps to reproduce the bug.
- [x] I've pasted the output of `argocd version`.

**Describe the bug**

After upgrading to v3.3.2, the application controller becomes CPU saturated and eventually OOMs. Goroutine profiles show 529 goroutines blocked on the `listSemaphore` and 440 blocked waiting for HTTP connections. Neither of which occurs in v2.14.15 under the same workload. 

I was blocked from updating to earlier 3.X versions because of #25517. 

We are seeing memory explosion after the clusters are fully synced. A similar issue as #25517. 

Flat (MB) | Flat% | Cum (MB) | Function
-- | -- | -- | --
5,820 | 14.9% | 11,751 | sigs.k8s.io/json.(*decodeState).objectInterface
5,292 | 13.5% | 5,329 | sigs.k8s.io/json.unquote
4,986 | 12.7% | 4,986 | bytes.growSlice
3,959 | 10.1% | 3,959 | reflect.New
1,696 | 4.3% | 3,635 | k8s.io/api/core/v1.(*Secret).DeepCopyInto
1,551 | 4.0% | 1,554 | k8s.io/client-go/tools/cache.(*threadSafeMap).Index
1,509 | 3.9% | 1,509 | argoproj/argo-cd.(*TLSClientConfig).DeepCopyInto
1,420 | 3.6% | 2,651 | k8s.io/apimachinery.(*ObjectMeta).DeepCopyInto
1,239 | 3.2% | 2,707 | argoproj/argo-cd.(*SettingsManager).updateSettingsFromSecret
97 | 0.25% | 8,602 | argoproj/argo-cd.(*SettingsManager).GetSettings (cumulative)
0 | — | 10,418 | argoproj/argo-cd.(*clusterInfoUpdater).updateClusters.func2 (cumulative)
0 | — | 10,457 | argoproj/argo-cd/util/argo.GetDestinationCluster (cumulative)

CPU taken over by GCing 
Controller | Duration | Total Samples | Effective CPU% | Top Flat Consumer
-- | -- | -- | -- | --
0 | 300.2s | 5,768.7s | 1921% | runtime.scanobject 5.6%
1 | 300.2s | 5,767.3s | 1921% | runtime/pprof.lostProfileEvent 9.7%
2 | 300.2s | 5,007.6s | 1668% | runtime.scanobject 6.6%
3 | 300.2s | 5,652.3s | 1883% | runtime.scanobject 5.3%

#26863 Fixes the below, thanks @jcogilvie  for the help there. 
**v3.3.2 (131,982 goroutines):**

| Count | Blocked on | Location |
|---|---|---|
| **529** | `semaphore.Acquire` | `listResources`  waiting for list semaphore |
| **440** | `Transport.getConn` | HTTP connection pool exhausted |
| 0 | `RWMutex.Lock` | No lock contention on `c.lock` |

**v2.14.15 (147,332 goroutines):**

| Count | Blocked on | Location |
|---|---|---|
| **0** | `semaphore.Acquire` | No semaphore saturation |
| **~0** | `Transport.getConn` | No HTTP connection saturation |
| 194 | `RWMutex.Lock` | Normal event processing + watch startup |

**CPU**
| Function | v2.14.15 (cumulative) | v3.3.2 (cumulative) | Delta |
|---|---|---|---|
| `newResource` | 24.4% | **27.3%** | **+22%** per-resource cost increase |
| `listResources` | 21.4% | **25.0%** | **+17%** listing path slower |
| `CompareAppState` | **29.6%** | 17.0% | Reconcile workers starved of CPU |
| `updateClusterInfo` | 1.07% | **19.26%** | 18× increase but expected because of #25534? |

~~I believe I have traced the regression to commit `ddce93cfd` (#24847, fixes #24379) but will need to add some profiling to our controller to get a better picture.~~

~~One thing that I noticed but haven't verified entirely yet is:~~

~~In `sync()`, every resource passes through:
listResources → EachListItem → newResource → [setNode](https://github.com/argoproj/argo-cd/blob/v3.3.2/gitops-engine/pkg/cache/cluster.go#L475) → [updateParentUIDToChildren](https://github.com/argoproj/argo-cd/blob/v3.3.2/gitops-engine/pkg/cache/cluster.go#L491)~~

~~After all resources are loaded, [sync() calls rebuildParentToChildrenIndex()](https://github.com/argoproj/argo-cd/blob/v3.3.2/gitops-engine/pkg/cache/cluster.go#L1108), which clears the entire index and rebuilds it from scratch. It feels like `rebuildParentToChildrenIndex` is redundant which would be O(total resources) of wasted work per cluster sync.~~

**To Reproduce**

**Expected behavior**

**Screenshots**

**Version**

```shell
Paste the output from `argocd version` here.
```

**Logs**

```
Paste any relevant application logs here.
```

### PR DESCRIPTION
**Pull request by jcogilvie**
Fixes #26863.

This PR contains two changes:

1. switch the parent-to-child index from a map of slices to a map of sets, eliminating O(k) scans for add and remove of children:  now O(1)
2. process index updates in-band when handling inferred refs, eliminating two problems:
  a. the performance impact of rebuilding the index on every sync (now obviated since the index is fully maintained in-band)
  b. a newly-discovered latent issue where index updates relating to inferred refs on watch events were missed

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by jcogilvie** [comment]
# Updated Benchmark Results

## Summary of Optimizations

This PR now includes **two optimizations**:

1. **Set-based index** - O(1) add/remove instead of O(k) slice scanning
2. **Inline index updates** - Eliminated redundant `rebuildParentToChildrenIndex()` call after sync

---

## Add/Remove Operations (O(1))

| Children per Parent | Time (ns/op) | Allocations |
|---------------------|--------------|-------------|
| 10 | 66.36 | 0 |
| 50 | 66.83 | 0 |
| 100 | 66.72 | 0 |
| 500 | 66.67 | 0 |
| 1000 | 68.32 | 0 |

Time is constant (~67 ns) regardless of children count, confirming O(1) behavior.

### Comparison with Original Slice-based Implementation

| Children | Original (slice, O(k)) | New (set, O(1)) | Speedup |
|----------|------------------------|-----------------|---------|
| 10 | ~115 ns | 67 ns | 1.7× |
| 50 | ~329 ns | 67 ns | 5× |
| 100 | ~581 ns | 71 ns | 8× |
| 500 | ~2,600 ns | 67 ns | 39× |
| 1000 | ~5,100 ns | 68 ns | **75×** |

---

## Rebuild Elimination

Previously, `sync()` would:
1. Call `setNode()` for each resource → incremental index updates
2. Call `rebuildParentToChildrenIndex()` → clear and rebuild entire index (O(n))

**The rebuild is now eliminated.** The index is maintained purely through inline updates:
- Explicit owner refs → updated in `updateParentUIDToChildren()`
- Inferred owner refs → updated inline in `setNode()` when detected

### Impact on Sync Performance

The rebuild overhead scaled linearly with resource count:

| Resources | Rebuild Time (eliminated) |
|-----------|---------------------------|
| 1,000 | ~0.1ms |
| 5,000 | ~0.6ms |
| 10,000 | ~1.2ms |
| 20,000 | ~2.7ms |

This overhead is **now zero** since the rebuild is no longer called.

---

## Incremental Index Build Performance

| Scenario | Time (ns/op) | Allocations |
|----------|--------------|-------------|
| 100 parents × 10 children | 115,436 | 511 |
| 100 parents × 50 children | 551,105 | 911 |
| 100 parents × 100 children | 1,039,556 | 1,111 |
| 1000 parents × 10 children | 1,150,421 | 5,022 |
| 1000 parents × 100 children | 9,533,911 | 11,022 |

---

## Additional Fix: Watch Event Bug

The inline index updates also fix a pre-existing bug where **inferred owner refs added via watch events** were not captured in the index.

| Scenario | Before | After |
|----------|--------|-------|
| Initial sync + explicit ref | ✓ | ✓ |
| Initial sync + inferred ref | ✓ (via rebuild) | ✓ (inline) |
| Watch event + explicit ref | ✓ | ✓ |
| Watch event + inferred ref | **✗ BUG** | ✓ (inline) |

New test `TestStatefulSetPVC_WatchEvent_IndexUpdated` verifies this fix.

---

## Memory Impact

| Children per parent | Slice | Set | Overhead |
|---------------------|-------|-----|----------|
| 10 | 664B | 848B | +28% |
| 100 | 6.4KB | 8KB | +25% |

**Cluster-level estimate** (1,000 parents × 10 children = 10,000 relationships):
- Additional overhead: **~18KB**

File: gitops-engine/pkg/cache/cluster_test.go [hunk 1]
```diff
@@ -2189,3 +2189,369 @@ func TestIterateHierarchyV2_NoDuplicatesCrossNamespace(t *testing.T) {
...
 	assert.Equal(t, 1, visitCount["cluster-child"], "cluster child should be visited once")
 }
+
+// BenchmarkSync_ParentToChildrenIndex measures the overhead of parent-to-children index
+// operations during sync. This benchmark was created to investigate performance regression
+// reported in https://github.com/argoproj/argo-cd/issues/26863
+//
+// The concern is that during sync():
+// 1. setNode() calls updateParentUIDToChildren() for each resource - O(n)
+// 2. After all resources are loaded, rebuildParentToChildrenIndex() is called - O(n) again
```
Comment (rumstead) [inline gitops-engine/pkg/cache/cluster_test.go]: Why do we have to do this twice? once for set node and then again in `rebuildParentToChildrenIndex`? 

In sync(), every resource passes through:
listResources → EachListItem → newResource → [setNode](https://github.com/argoproj/argo-cd/blob/v3.3.2/gitops-engine/pkg/cache/cluster.go#L475) → [updateParentUIDToChildren](https://github.com/argoproj/argo-cd/blob/v3.3.2/gitops-engine/pkg/cache/cluster.go#L491)

After all resources are loaded, [sync() calls rebuildParentToChildrenIndex()](https://github.com/argoproj/argo-cd/blob/v3.3.2/gitops-engine/pkg/cache/cluster.go#L1108), which clears the entire index and rebuilds it from scratch.

Comment (rumstead) [reply gitops-engine/pkg/cache/cluster_test.go]: I am fairly confident we can remove `rebuildParentToChildrenIndex` entirely.

Comment (jcogilvie) [reply gitops-engine/pkg/cache/cluster_test.go]: there are a few factors at play here that made both the incremental update and the bulk rebuild necessary for different conditions (which i have now fixed):  

1.  setNode is used in two paths:  initial sync and watch events.  in the initial sync, the bulk update was necessary to capture inferred refs.  for watch events, the bulk update never happens, so we relied on the incremental updates (and potentially missed updates to inferred owners -- this is a separate bug which my fix now addresses).  
2. we did not want to track "are we syncing?" as a state here because it introduces the potential for race conditions when watch events come in during a sync

after addressing the inferred refs factor, we can now eliminate the need for the bulk rebuilds -- which will also benefit from the slice -> set conversion i've done.

File: gitops-engine/pkg/cache/cluster_test.go [hunk 2]
```diff
@@ -2190,111 +2307,231 @@ func TestIterateHierarchyV2_NoDuplicatesCrossNamespace(t *testing.T) {
...
-				UID:        "self-ref-uid", // Points to itself
-			}},
-		},
+// BenchmarkSync_ParentToChildrenIndex measures the overhead of parent-to-children index
+// operations during sync. This benchmark was created to investigate performance regression
+// reported in https://github.com/argoproj/argo-cd/issues/26863
+//
+// The index is now maintained with O(1) operations (set-based) and updated inline
+// in setNode() for both explicit and inferred owner refs. No rebuild is needed.
+//
```
[the remark this answers is not in the document]
Comment (jcogilvie) [reply gitops-engine/pkg/cache/cluster_test.go:2430]: Restored.

File: gitops-engine/pkg/cache/cluster.go [hunk 3]
```diff
@@ -504,60 +505,65 @@ func (c *clusterCache) setNode(n *Resource) {
...
-		}
-	}
-	c.parentUIDToChildren[parentUID] = append(children, childKey)
+	// Get or create the set for this parent
+	childrenSet := c.parentUIDToChildren[parentUID]
+	if childrenSet == nil {
+		childrenSet = make(map[kube.ResourceKey]struct{})
+		c.parentUIDToChildren[parentUID] = childrenSet
+	}
+	// Add child to set (O(1) operation, automatically handles duplicates)
```
Comment (rumstead) [inline gitops-engine/pkg/cache/cluster.go:550]: a super nit but i think we can remove this comment.

File: gitops-engine/pkg/cache/cluster_test.go [hunk 4]
```diff
@@ -2298,3 +2420,226 @@ func TestIterateHierarchyV2_CircularOwnerChain_NoStackOverflow(t *testing.T) {
...
+
+			for n := 0; n < b.N; n++ {
+				// Clear the index
+				cluster.parentUIDToChildren = make(map[types.UID]map[kube.ResourceKey]struct{})
+
+				// Simulate incremental adds (O(1) set insertions)
+				for _, child := range children {
+					cluster.addToParentUIDToChildren(child.parentUID, child.childKey)
+				}
+			}
```
[the remark this answers is not in the document]
Comment (rumstead) [reply gitops-engine/pkg/cache/cluster_test.go:2642]: This feels like a good suggestion?

**Review by rumstead — APPROVED** [review]
From a performance point of view, this PR addresses a lot of what I saw. Since the tests are passing I am assuming that functionally this works as well. 

LGTM

**Comment by rumstead** [comment]
I am opting to cherry pick into 3.3 and 3.4 but open to opinions if we shouldn't. I plan on pulling in these changes and testing out a patch version with some addition logging as well in the next week or so.

**Comment by agaudreault** [comment]
@rumstead have you tried this patch in your environment and confirmed it fixes #26863 ? If not, do you think it is something you could do before merging this? 

FYI, I dont see any problems merging this, but I want to confirm it fixes the issue.

**Comment by leoluz** [comment]
@rumstead were you able to validate 3.2 to confirm it it has the issue as well? We briefly discussed this during contrib meeting today and we just wanted to doublecheck that 3.2 also has the issue before moving forward with the cherry-picking.

**Comment by rumstead** [comment]
@agaudreault yes we are planning next week to test the patch. We should probably merge this regardless of the results. It does fix an edge case and removes some redundant work on the `sync` hot path. 

@leoluz 3.2 doesn't have this change and should be unaffected. It was [introduced only in 3.3](https://github.com/argoproj/argo-cd/commits/release-3.3/?since=2025-12-14&until=2025-12-15) and isn't in [any 3.2 tag](https://github.com/argoproj/argo-cd/commit/ddce93cfdd5d23769d0b2e19f0887775d1110b40). I only set the cherry-pick labels for 3.3 and 3.4 which are impacted.

**Review by leoluz — APPROVED** [review]

**Comment by rumstead** [comment]
See the issue for more details but this change helped!
